mlflow/llama_index/tracer.py

```diff
@@ -0,0 +1,295 @@
+        so the instance metadata serves as a fallback source of information.
+        """
+
+    # TODO: Union type hint doesn't work with singledispatchmethod, so we have to define
+    #  two separate methods for BaseLLM and MultiModalLLM. Once we upgrade to Python 3.10,
+    #  we can use `BaseLLM | MultiModelLLM` type hint and it works with singledispatchmethod.
+    @_get_instance_attributes.register
+    def _(self, instance: BaseLLM):
+        return self._get_llm_attributes(instance)
+
+    @_get_instance_attributes.register
+    def _(self, instance: MultiModalLLM):
+        return self._get_llm_attributes(instance)
+
+    def _get_llm_attributes(self, instance) -> Dict[str, Any]:
+        attr = {}
+        if metadata := instance.metadata:
+            attr["model_name"] = metadata.model_name
+            if params_str := metadata.json(exclude_unset=True):
+                attr["invocation_params"] = json.loads(params_str)
+        return attr
+
+    @_get_instance_attributes.register
+    def _(self, instance: BaseEmbedding):
+        return {
+            "model_name": instance.model_name,
+            "embed_batch_size": instance.embed_batch_size,
+        }
+
+    @_get_instance_attributes.register
+    def _(self, instance: BaseTool):
+        metadata = instance.metadata
+        attributes = {"description": metadata.description}
+        try:
+            attributes["name"] = metadata.name
+        except ValueError:
+            # ToolMetadata.get_name() raises ValueError if name is None
+            pass
+        try:
+            attributes["parameters"] = json.loads(metadata.fn_schema_str)
+        except ValueError:
+            # ToolMetadata.get_fn_schema_str() raises ValueError if fn_schema is None
+            pass
+        return attributes
+
+
+class MlflowEventHandler(BaseEventHandler, extra="allow"):
+    """
+    Event handler processes various events that are triggered during execution.
+
+    Events are used as supplemental source for recording additional metadata to the span,
+    such as model name, parameters to the span, because they are not available in the inputs
+    and outputs in SpanHandler.
+    """
+
+    _span_handler: MlflowSpanHandler
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "MlflowEventHandler"
+
+    def __init__(self, _span_handler):
+        super().__init__()
+        self._span_handler = _span_handler
+
+    def handle(self, event: BaseEvent) -> Any:
+        # TODO: These two InProgress events are triggered per streaming chunk. We
+        # ignore them for now but they should be accumulated as SpanEvents of the parent span.
+        if isinstance(event, (LLMChatInProgressEvent, LLMCompletionInProgressEvent)):
+            return
+
+        if span := self._span_handler.open_spans.get(event.span_id):
+            self._handle_event(event, span._mlflow_span)
+
+    @singledispatchmethod
+    def _handle_event(self, event: BaseEvent, span: LiveSpan):
+        ...
```

Comment:
```suggestion
        raise NotImplementedError
```

should we raise `NotImplementedError` here?

Reply:
We don't handle all events, this is no-op impl for events we are not interested in.

Reply:
Can we use `pass`? `...` looks like "forgot to update the code.".

Reply:
Makes sense!